ntcore/src/main/java/edu/wpi/first/networktables/ConnectionListener.java

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package edu.wpi.first.networktables;
+
+import edu.wpi.first.util.WPIUtilJNI;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Consumer;
+
+/** Connection listener. This calls back to a callback function when a connection change occurs. */
+public final class ConnectionListener implements AutoCloseable {
+  /**
+   * Create a listener for connection changes.
+   *
+   * @param inst Instance
+   * @param immediateNotify if notification should be immediately created for existing connections
+   * @param listener Listener function
+   */
+  public ConnectionListener(
+      NetworkTableInstance inst,
+      boolean immediateNotify,
+      Consumer<ConnectionNotification> listener) {
+    s_lock.lock();
+    try {
+      if (s_poller == 0) {
+        s_inst = inst;
+        s_poller = NetworkTablesJNI.createConnectionListenerPoller(inst.getHandle());
+        startThread();
+      }
+      m_handle = NetworkTablesJNI.addPolledConnectionListener(s_poller, immediateNotify);
+      s_listeners.put(m_handle, listener);
+    } finally {
+      s_lock.unlock();
+    }
+  }
+
+  @Override
+  public synchronized void close() {
+    if (m_handle != 0) {
+      s_lock.lock();
+      try {
+        s_listeners.remove(m_handle);
+      } finally {
+        s_lock.unlock();
+      }
+      NetworkTablesJNI.removeConnectionListener(m_handle);
+      m_handle = 0;
+    }
+  }
+
+  /**
+   * Determines if the native handle is valid.
+   *
+   * @return True if the native handle is valid, false otherwise.
+   */
+  public boolean isValid() {
+    return m_handle != 0;
+  }
+
+  private int m_handle;
+
+  private static final ReentrantLock s_lock = new ReentrantLock();
+  private static final Map<Integer, Consumer<ConnectionNotification>> s_listeners = new HashMap<>();
+  private static Thread s_thread;
+  private static NetworkTableInstance s_inst;
+  private static int s_poller;
+  private static boolean s_waitQueue;
+  private static final Condition s_waitQueueCond = s_lock.newCondition();
+
+  private static void startThread() {
```

Comment:
I'm seeing a lot of repetition between the `*Listener` and `*ListenerPoller` classes. Any point in trying to extract common superclasses or something?

Reply:
I don’t know if it’s possible to extract, it’s all different underlying types and JNI calls.